README
NAME
    Net::DNS::ToolKit - tools for working with DNS packets

SYNOPSIS
      use Net::DNS::ToolKit qw(

            get1char
            get16
            get32
            put1char
            put16
            put32
            getIPv4
            putIPv4
            putIPv6
            getIPv6
            getstring
            putstring
            dn_comp
            dn_expand
            parse_char
            gethead
            newhead
            getflags
            putflags
            get_qdcount
            get_ancount
            get_nscount
            get_arcount
            put_qdcount
            put_ancount
            put_nscount
            put_arcount
            inet_aton
            inet_ntoa
            ipv6_aton
            ipv6_n2x
            ipv6_n2d   
            sec2time
            ttlAlpha2Num
            collapse
            strip
            get_ns
            gettimeofday
      );

      $char = get1char(\$buffer,$offset);
      ($int, $newoff)  = get16(\$buffer,$offset);
      ($long, $newoff) = get32(\$buffer,$offset);
      $newoff = put1char(\$buffer,$offset,$u_char);
      $newoff = put16(\$buffer,$offset,$int);
      $newoff = put32(\$buffer,$offset,$long);
      $flags = getflags(\$buffer);
      true = putflags(\$buffer,$flags);
      $int = get_qdcount(\$buffer);
      $int = get_ancount(\$buffer);
      $int = get_nscount(\$buffer);
      $int = get_arcount(\$buffer);
      $newoff = put_qdcount(\$buffer,$int);
      $newoff = put_ancount(\$buffer,$int);
      $newoff = put_nscount(\$buffer,$int);
      $newoff = put_arcount(\$buffer,$int);
      ($netaddr,$newoff)=getIPv4(\$buffer,$offset);
      $newoff = putIPv4(\$buffer,$offset,$netaddr);
      ($ipv6addr,$newoff)=getIPv6(\$buffer,$offset);
      $newoff = putIPv6(\$buffer,$offset,$ipv6addr);
      ($offset,
       $id,$qr,$opcode,$aa,$tc,$rd,$ra,$mbz,$ad,$cd,$rcode,
       $qdcount,$ancount,$nscount,$arcount)
            = gethead(\$buffer);
      $newoff = newhead(\$buffer,$id,$flags,
            $qdcount,$ancount,$nscount,$arcount);
      ($b,$h,$d,$a)=parse_char($char);
      ($newoff,$name) = dn_expand(\$buffer,$offset);
      ($newoff,@dnptrs)=dn_comp(\$buffer,$offset,\$name,\@dnptrs);
      $dotquad = inet_ntoa($netaddr);
      $netaddr = inet_aton($dotquad);
      $ipv6addr = ipv6_aton($ipv6_text);
      $hex_text = ipv6_n2x($ipv6addr);
      $dec_text = ipv6_n2d($ipv6addr);
      $timetxt = sec2time($seconds);
      $seconds = ttlAlpha2Num($timetext);
      $shorthost = collapse($zonename,$longhost);
      $tag = strip($P_tag);
      @nameservers = get_ns();
      ($secs,$usecs) = gettimeofday();

DESCRIPTION
    Routines to pick apart, examine and put together DNS packets. They can
    be used for diagnostic purposes or as building blocks for DNS
    applications such as DNS servers and clients or to allow user
    applications to interact directly with remote DNS servers.

      See: Net::DNS::ToolKit:RR and the subdirectory
            lib/Net/DNS/ToolKit/RR/
      for individual Resource Record methods.

      Net::DNS::ToolKit does not handle every type of RR with context
      help for the record format. HOWEVER, it does handle all unknown
      record types per RFC-3597 so if your program can manipulate the
      binary and/or hex representation of the data as proscribed in RFC-3597 this
      module will always work for you.

    A good example of full utilization of this module is the Net::DNS::Dig
    manpage/module.

    See: the Net::DNS::ToolKit::RR manpage (included in this distribution)
    for a complete description of how to use this module and the
    accompanying Resource Records tools.

FUNCTIONS
    These functions return a value and offset in list context and first
    value only in scalar context.

      ($int,$newoff)        = get16(...
      ($long,$newoff)       = get32(...
      ($netaddr,$newoff)    = getIPv4(...
      ($ipv6addr,$newoff)   = getIPv6(...
      ($string,$newoff)     = getstring(...
      ($newoff,$name)       = dn_expand(...
      ($secs,$usecs)        = gettimeofday(...

    These functions return only a value or an offset.

      $newoff       = put1char(...
      $newoff       = put16(...  
      $newoff       = put32(...
      $newoff       = put_qdcount(...
      $newoff       = put_ancount(...
      $newoff       = put_nscount(...
      $newoff       = put_arcount(...
      $newoff       = putIPv4(...
      $newoff       = putIPv4(...
      $newoff       = putstring(...
      $newoff       = newhead(...
      $flags        = getflags(...
      true          = putflags(...
      $int          = get_qdcount(...
      $int          = get_ancount(...
      $int          = get_nscount(...
      $int          = get_arcount(...
      $char         = get1char(...
      $dotquad      = inet_ntoa(...
      $netaddr      = inet_aton(...
      $timetxt      = sec2time(...
      $seconds      = ttlAlpha2Num(...
      $tag          = strip(...
      $shorthost    = collapse(...

    This function always return list context prefixed by a new offset.

      ($newoff,@dnptrs) = dn_comp(...
      ($offset,@list)   = gethead(...

    These functions always return list context.

      @list         = parse_char(...
      @nameservers  = get_ns(...

    * $char = get1char(\$buffer,$offset);
        Get a single character from the buffer at $offset

          input:        pointer to buffer,
                        offset into buffer
          output:       the "character"   
                   or   undef if the pointer
                        is outside the buffer

    * ($int, $newoff) = get16(\$buffer,$offset);
        Get a 16 bit integer from the buffer at $offset. Return the value
        and a new offset pointing at the next character.

        Returns and empty array on error.

          input:        pointer to buffer,
                        offset into buffer
          returns:      16 bit integer,
                        offset + size of int

        In SCALAR context, returns just the value.

    * $newoff = put1char(\$buffer,$offset,$u_char);
        Put an unsigned 8 bit value into the buffer at $offset. Return the
        value of the new offset pointer to the next char (usually end of
        buffer).

    * $newoff = put16(\$buffer,$offset,$int);
        Put a 16 bit integer into the buffer at $offset. Return the value of
        the new offset pointer to the the next char (usually end of buffer).

          input:        pointer to buffer,
                        offset into buffer,
                        16 bit integer
          returns:      offset + size of int

    * ($long, $newoff) = get32(\$buffer,$offset);
        Get a 32 bit long from the buffer at $offset. Return the long and a
        new offset pointing at the next character.

        Returns and empty array on error.

          input:        pointer to buffer,
                        offset into buffer
          returns:      32 bit long,
                        offset + size long

        In SCALAR context, returns just the value.

    * $newoff = put32(\$buffer,$offset,$long);
        Put a 32 bit long into the buffer at $offset. Return the value of
        the new offset pointer to the the next char (usually end of buffer).

          input:        pointer to buffer,
                        offset into buffer,
                        32 bit long
          returns:      offset + size of int

    * $flags = getflags(\$buffer);
        Get the flag bits from the header

          input:        pointer to buffer,
          returns:      flag bits

    * putflags(\$buffer,$flags);
        Put flags bits back in header

          input:        pointer to buffer,
                        flags bits
          returns:      n/a

    * $int = get_qdcount(\$buffer);
        Get the contents of the qdcount.

          input:        pointer to buffer,
          returns:      16 bit integer,

    * $int = get_ancount(\$buffer);
        Get the contents of the ancount.

          input:        pointer to buffer,
          returns:      16 bit integer,

    * $int = get_nscount(\$buffer);
        Get the contents of the nscount.

          input:        pointer to buffer,
          returns:      16 bit integer,

    * $int = get_arcount(\$buffer);
        Get the contents of the arcount.

          input:        pointer to buffer,
          returns:      16 bit integer,

    * $newoff = put_qdcount(\$buffer,$int);
        Put a 16 bit integer into qdcount. Return an offset to ancount.

          input:        pointer to buffer,
                        16 bit integer,
          returns:      offset to ancount

    * $newoff = put_ancount(\$buffer,$int);
        Put a 16 bit integer into ancount. Return an offset to nscount.

          input:        pointer to buffer,
                        16 bit integer,
          returns:      offset to nscount

    * $newoff = put_nscount(\$buffer,$int);
        Put a 16 bit chunk into nscount. Return an offset to arcount.

          input:        pointer to buffer,
                        16 bit integer,
          returns:      offset to arcount

    * $newoff = put_arcount(\$buffer,$int);
        Put a 16 bit integer into arcount. Return an offset to answer
        section.

          input:        pointer to buffer,
                        16 bit integer,
          returns:      offset to question section

    * ($netaddr,$newoff)=getIPv4(\$buffer,$offset);
        Get an IPv4 network address from the buffer at $offset. Return the
        netaddr and a new offset pointing at the next character beyond.

        Returns and empty array on error.

          input:        pointer to buffer,
                        offset into buffer
          returns:      netaddr,
                        offset + size of ipaddr

        In SCALAR context, returns just netaddr.

    * $newoff = putIPv4(\$buffer,$offset,$netaddr);
        Put a netaddr into the buffer. Return the value of the new offset
        pointer to the next char (usually end of buffer).

          input:        pointer to buffer,
                        offset into buffer,
                        packed IPv4 net address

Changes
Revision history for Perl extension Net::DNS::ToolKit

0.47  Tue Apr 16 11:59:02 PDT 2013
	add #ifdef BIND_4_COMPAT
	around #include nameser_compat.h
	so that later versions of OpenBSD work properly

0.46  Sun Apr  7 14:16:12 PDT 2013
	add support for SRV

0.45  Nov 17 11:52:05 PST 2011
	update buggy test for Windows OS in Makefile.pl, bug #72478
	Thanks to Chrisv for spotting that.

0.44  Sat Nov  5 14:02:42 PDT 2011
	add support for 'unknown RR's' per RFC 3597
	add support for 
		AFSDB
		ISDN
		MD
		MF
		RT
		X25

0.43  Fri Sep 23 23:08:24 PDT 2011
	add test for OS "windows" and exit 0

	add "putflags" function

0.42  Sat Mar  7 17:38:03 PST 2009
	edits to xs_include/dn_expand.inc and ToolKit.xs
	to clean up char/u_char mismatches that "cc" cares about

0.41  Wed Oct 15 14:19:29 PDT 2008
	require Net::DNS::Codes version 0.09 or higher

0.40  Wed Oct 15 13:30:22 PDT 2008
	revised get_ns and replaced it with a cluster of XS and perl
	routines to circumvent thread issues in res_init and res_ninit

0.39  Mon Oct  6 13:16:02 PDT 2008
	clean up windows build bug

0.38  Thu Oct  2 14:49:26 PDT 2008
	update ToolKit.xs get_ns to be thread safe

0.37  Sat Sep 27 11:16:29 PDT 2008
	Include definition of MAXDNAME missing in osx
	Thanks to schneecrash@gmail.com for pointing this out

0.36  Wed Feb 27 11:12:58 PST 2008
	Fixed typo in sub 'newhead' where the wrong variable
	referenced in argument testing.
	Thanks to Rafal Krypa for spotting and reporting that error.

0.35  release date missing 
	update v0.05 of Utilities::dns_ptr to
	return a scalar or an array of names

0.34  Thu Oct  4 09:54:14 PDT 2007
	include <arpa/nameser_compat.h>
	so build works on OSX
	Thanks to Chris Turner chris@turnerhq.com for providing
	the bug fix.

0.33  Wed Sep 19 21:28:00 PDT 2007
	correct typo in variable name in dig.pl

0.32  Wed Sep 19 21:08:21 PDT 2007
	add "print header" switch to dig.pl

0.31  Wed Apr 11 11:43:23 PDT 2007
	RR.pm ver 0.06, small documentation update

	Thanks to "Clayton O'Neill" <coneill@oneill.net> for spotting this
	one. Fixed bug in dn_comp which only seems to manifest when you pass in
	a name that begins with a number.  Sometimes it seems that if perl has
	evaluated the scalar in a numeric context for some reason, it's type
	will be considered to be IV, not PV.  The patch I've attached changes
	the check to see if it's a glob first, and if not, checks to see if it
	can be used as a PV, instead of checking if it is already a PV.

0.30  Thu Oct 19 09:43:36 PDT 2006
	Added RP and MINFO types

0.29  Mon Jul 24 14:07:10 PDT 2006
	remove unused reference to NetAddr::IP in
	Net::DNS::ToolKit::Utilities

0.28  Mon Jun  5 15:04:20 PDT 2006
	remove references to miniSocket.inc in Makefile.PL
	this is now imported from NetAddr/IP/Util

	fix solaris compiler ENDIANness problems

0.27  Tue Mar 28 21:02:54 PST 2006
	migrate inex_xxx and ipv6_xxx functions to 
	NetAddr::IP::Util

0.26  Tue Mar 21 19:44:57 PST 2006
	update MX.pm to correct documentation typo

	fix ipv6_aton to correctly convert pathological case of
	::1:2:3:4:5:6:7 and 1:2:3:4:5:6:7:: where there are 8 colons
	i.e. ::nnn... ...nnn:: 	instead of starting or ending with 
	:nnn nnn: or 0:nnn... ...nnn:0

0.25  Sat Jun 11 23:34:32 PDT 2005
	'new' failed to build a packet with a transaction ID of '0'
	Thanks to "Clayton O'Neill" <coneill@oneill.net> for 
	spotting that.

0.24  Tue Feb 22 21:47:21 PST 2005
	add IPv6 support for return records (not query)
		ipv6_aton
		ipv6_n2x
		ipv6_n2d
		getIPv6  
		putIPv6

0.23  Sat Nov 27 17:08:45 PST 2004
	removed a legacy package 'use package'
	that is not called in Net::DNS::ToolKit::Utilities

0.22  Sun Nov 14 23:33:19 PST 2004
	corrected a 'use package' call in t/ss_id.t
	call should be to Net::DNS::ToolKit::Utilities
	and was to a legacy package that is not a dependency
	of this module

0.21  Sat Sep 18 12:33:59 PDT 2004
	update docs in Utilities.pm

0.20  Wed Sep  1 15:56:25 PDT 2004
	Create Utilities.pm from portions of
	Mail::SpamCannibal::ScriptSupport

	see comments in Mail::SpamCannibal v0.44 Changes

0.19  not released
	update rdns_blk.pl v0.03, add optional -d switch
	to show rDNS response for bad lookups

0.18  Thu Jun 24 15:38:36 PDT 2004
	update rdns_blk.pl v0.02 to print PTR when response is a C_NAME

0.17  Fri Jun  4 13:23:16 PDT 2004
	correct documentation error for "parse_char"
	correct documentation error in Net::DNS::ToolKit::RR::CNAME v0.04

	add script 'rdns_blk.pl' to do lookups on cidr24 reverse IP's

0.16  Sat Nov 22 14:32:27 PST 2003
	There is a bug that I can't reproduce here on either the 
	linux or sun boxes for version of perl 5.003 -> 5.8x so I've
	added a trap in get_ns.t to check hosts for problems with 
	resolver not returning the local host name server IP addy's
	Hopefully, someone will find such a host and provide a user
	account so I can find the problem and fix it.

	DOCUMENTATION change only.
	correct the comments for $put->CNAME,MB,MG,MR,NS,PTR
	erronously said that $rdlength was an input ARG when
	it is actually calculated and inserted by the function

0.15  Sat Nov  1 09:44:17 PST 2003

lib/Net/DNS/ToolKit/RR.pm
package Net::DNS::ToolKit::RR;

#use 5.006;
use strict;
#use diagnostics;
#use warnings;

use Net::DNS::Codes qw(:RRs);
use Net::DNS::ToolKit qw(
	get16
	get32
	put16
	put32
	getstring
	dn_comp
	dn_expand
);
use vars qw($VERSION $autoload *sub);
require Net::DNS::ToolKit::Question;

$VERSION = do { my @r = (q$Revision: 0.07 $ =~ /\d+/g); sprintf "%d."."%02d" x $#r, @r };

sub remoteload {
#    *sub = $autoload;
    (my $RRtype = $autoload ) =~ s/.*::(\w+):://;
    # function = $1, one of get,put,parse
    ($autoload,$_) = instantiate($RRtype,$1);
#    my $code = 'package '. __PACKAGE__ .'::'. $1 .'; '.'*'. $RRtype .'=\&'. $autoload;
    my $code = 'package '. __PACKAGE__ .'::'. $1 .'; '.'*'. $RRtype .
	q| = sub { unshift @_,'|. $autoload . q|'; &|. $_ .'};';
    eval "$code";

# print "AUTOLOAD=",*sub,";\n";
# print "subname=$autoload RRtype=$RRtype func=$1\n";
# print 'code=', $code, "\n";

#    no strict;
#    eval { *sub = sub { unshift @_,$autoload; &$_ } };
#    goto &{*sub};
    unshift @_,$autoload;
    goto &$_;
}

# return target function, target interpreter
sub instantiate {
    my($RRtype,$func) = @_;
    if ($RRtype eq 'DESTROY') {	# should never get here
	die __PACKAGE__.".pm: DESTROY must be defined internally in the calling package\n";
    } else {
	my $filename = __PACKAGE__.'::'.$RRtype.'.pm';
	$filename =~ s#::#/#g;
	my $save = $@;
	eval { local $SIG{__DIE__}; require $filename };
	if ($@) {
#	    die __PACKAGE__.'::RR'.$func.' not implemented'
#		if $func eq 'put';
#	    $@ = $save;
#	    $RRtype = 'NotImplemented';
	  my $generic;
	  if (	$RRtype =~ /^TYPE(\d+)$/ &&
		($generic = TypeTxt->{$1}) &&
		$generic =~ /T_(.+)/) {
		$generic = __PACKAGE__.'::'. $1;
	  } else {
	    $generic = __PACKAGE__.'::TYPE';
	  }
	  local $_ = $generic .'.pm';
	  s#::#/#g;
	  require $_;
	  my $code = 'package '. __PACKAGE__ .'::'. $RRtype .';
*get = \&'. $generic .'::get;
*put = \&'. $generic .'::put;
*parse = \&'. $generic .'::parse;';
	  eval "$code";
	}
    }
    # package from local scope
    return (__PACKAGE__.'::'.$RRtype.'::'.$func, __PACKAGE__.'::RR'.$func);
}

# return instantiated function
sub make_function {
  my $type = shift;
  (caller(1))[3] =~ /RR(\w+)$/;
  my $action = $1;
  if (($_ = TypeTxt->{$type}) && $_ =~ /T_(.+)/) {	# type is real?
    my $function = __PACKAGE__.'::'.$1;
    if ($function->can($action)) {	# if function is instantiated
      return $function .= '::'.$action;
    } else {				# instantiate it or NotImplemented
      return (instantiate($1,$action))[0];
    }
  } else {
#  return __PACKAGE__.'::NotImplemented::'.$action;
    my $function = __PACKAGE__.'::TYPE'. $type;
    if ($function->can($action)) {	# if function is instantiated
      return $function .= '::'.$action;
    } else {				# instantiate it or NotImplemented
      return (instantiate("TYPE$type",$action))[0];
    }
  }
}  

#########################################################
#	implements the common portion of...
#  ($newoff,$name,$type,$class,$ttl,$rdlength,$rdata,...)
#        = $get->next(\$buffer,$offset);

sub RRget {
  my($function,$self,$bp,$newoff) = @_;
  my ($off,$name) = dn_expand($bp,$newoff);
  (my $type, $off) = get16($bp,$off);
  (my $class, $off) = get16($bp,$off);
  (my $ttl, $off) = get32($bp,$off);
  my $rdlength = get16($bp,$off);	# scalar context, don't get offset
  $function = make_function($type) unless $function;
  no strict;
  ($off, my @results) = &$function($self,$bp,$off);
  return($off,$name,$type,$class,$ttl,$rdlength,@results);
}

#########################################################
#	implements the common portions of...
#  ($newoff,@dnptrs)=$put->XYZ(\$buffer,$offset,\@dnptrs,
#        $name,$type,$class,$ttl,$rdata,...);

sub RRput {
  # extract common elements from input, shrink input
  # input was: $function,$self,\$buffer,$offset,\@dnptrs,$name,$type,$class,$ttl,@rdata
  my ($func,$put,$bp,$off,$dnp,$name,$type,$class,$ttl) = @_;
  if (exists $_[1]->{class}) {
    ($func,$put,$bp,$off,$dnp,$name,$ttl) = splice(@_,0,7);
    $class = $put->{class};
    $func =~ /.+::(.+)::put$/;
    $type = 'T_'.$1;
    no strict;
    $type = &$type;
  } else {
    ($func,$put,$bp,$off,$dnp,$name,$type,$class,$ttl) = splice(@_,0,9);
  }
  # input is now: @rdata
  ($off, my @dnptrs) = dn_comp($bp,$off,\$name,$dnp);
  unless (@dnptrs) {		# if not valid return
    while(shift) {};		# empty the input array
    return ();			# error
  }
  return () unless ($off = put16($bp,$off,$type));
  # the rest should work since offset has been checked
  $off = put16($bp,$off,$class);# class
  $off = put32($bp,$off,$ttl);# ttl
  no strict;
  &$func($self,$bp,$off,\@dnptrs,@_);
}

####################################################################
#	implements the common portion of...
#  ($name,$typeTXT,$classTXT,$ttl,$rdlength,$RDATA,...)
#        = $parse->XYZ($name,$type,$class,$ttl,$rdlength,$rdata,...)

sub RRparse {
  # extract common elements from input, shrink input
  # input was: $function,$self,$name,$type,$class,$ttl,$rdlength,@rdata
  my $function = shift;
  # input is now: $name,$type,$class,$ttl,$rdlength,@rdata
  my ($name,$type,$class,$ttl,$rdlength) = splice(@_,1,5);	# pass $self,@rdata to $function call
# if length is ever needed, add it here
#  $_[0]->{len} = $rdlength;
  $name .= '.';	# terminate domain name
  $function = make_function($type) unless $function;
  no strict;
  my $typetxt = TypeTxt->{$type} || "TYPE$type";
  my $classtxt = ClassTxt->{$class} || "CLASS$class";
  return($name,$typetxt,$classtxt,$ttl,$rdlength,&{$function}(@_));
}

#####################################################################
######################### sub PACKAGES ##############################
#####################################################################

# this entire sub package is obsolete as of v0.07
#{
#  package Net::DNS::ToolKit::RR::NotImplemented;
#
#  sub get {
#    my($self,$bp,$offset) = @_;
#    (my $rdlength, $offset) = &Net::DNS::ToolKit::get16($bp,$offset);
#    $offset += $rdlength;
#    return($offset,"\0");
#  }
#
## die in loader, unimplemented
##  sub put {
##    my($bp,$off,$dp) = @_;
##    return($off,@$dp);
##  }
#  
#  sub parse {
#    shift;	# $self
#    return(@_);	# garbage in, garbage out
#  }
#}

{
    package Net::DNS::ToolKit::RR::get;
    use vars qw($AUTOLOAD);

    # preload Question
    *Question = \&Net::DNS::ToolKit::Question::get;

    sub AUTOLOAD {
	$Net::DNS::ToolKit::RR::autoload = $AUTOLOAD;
	goto &Net::DNS::ToolKit::RR::remoteload;
    }
    sub next {
      unshift @_,undef;	# flag to RRget;
      goto &Net::DNS::ToolKit::RR::RRget;
    }
    sub EmptyList {()};
    sub DESTROY {};
}

{
    package Net::DNS::ToolKit::RR::put;
    use vars qw($AUTOLOAD);

    # preload Question
    *Question = \&Net::DNS::ToolKit::Question::put;

    sub AUTOLOAD {
	$Net::DNS::ToolKit::RR::autoload = $AUTOLOAD;
	goto &Net::DNS::ToolKit::RR::remoteload;
    }
    sub DESTROY {};
}

{
    package Net::DNS::ToolKit::RR::parse;
    use vars qw($AUTOLOAD);

    # preload Question
    *Question = \&Net::DNS::ToolKit::Question::parse;

    sub AUTOLOAD {
	$Net::DNS::ToolKit::RR::autoload = $AUTOLOAD;
	goto &Net::DNS::ToolKit::RR::remoteload;
    }
# this next sub has been in the distro a long time
# $parse->RR
# this was unintentional but does not hurt anything
    sub RR {
      unshift @_,undef; # flag to RRparse; 
      goto &Net::DNS::ToolKit::RR::RRparse;
    }
# this SHOULD of been here instead of the above
    sub next {
      unshift @_,undef; # flag to RRparse; 
      goto &Net::DNS::ToolKit::RR::RRparse;
    }
    sub DESTROY {};
}

=head1 NAME

Net::DNS::ToolKit::RR - Resource Record class loader

=head1 SYNOPSIS

  use Net::DNS::ToolKit::RR;

  ($get,$put,$parse) = new Net::DNS::ToolKit::RR;
	or
  ($get,$put,$parse) = Net::DNS::ToolKit::RR->new;

	retrieve the next record (type unknown)
  ($newoff,$name,$type,$class,$ttl,$rdlength,$rdata,...)
	= $get->next(\$buffer,$offset);

	parse the current record (type in input fields)
  ($name,$typeTXT,$classTXT,$ttlTXT,$rdlength,$RDATA,...)
	= $parse->RR($name,$type,$class,$ttl,$rdlength,
			$rdata,...);

  ($newoff,@dnptrs)=$put->XYZ(\$buffer,$offset,\@dnptrs,
	$name,$type,$class,$ttl,$rdata,...);

  The 'get' and 'parse' operations can also be done
  by specific record type...
  ...but why would you use them instead of 'next' & 'RR'?

  ($newoff,$name,$type,$class,$ttl,$rdlength,$rdata,...)
	= $get->XYZ(\$buffer,$offset);

  ($name,$typeTXT,$classTXT,$ttlTXT,$rdlength,$RDATA,...)
	= $parse->XYZ($name,$type,$class,$ttl,$rdlength,
			$rdata,...);

	or you can use the individual methods 
	directly without calling "new"

  @output=Net::DNS::ToolKit::RR::get->next(@input);
  @output=Net::DNS::ToolKit::RR::get->XYZ(@input);
  @output=Net::DNS::ToolKit::RR::put->XYZ(@input);
  @output=Net::DNS::ToolKit::RR::parse->RR(@input);
  @output=Net::DNS::ToolKit::RR::parse->XYZ(@input);

The Question section is a special case:

  ($newoff,$name,type,class) = 
	$get->Question(\$buffer,$offset);
  ($newoff,@dnptrs) = 
	$put->Question(\$buffer,$offset,
	$name,$type,$class,\@dnptrs);
  ($name,$typeTXT,$classTXT) =
	$parse->Question($name,$type,$class);

=head1 ALTERNATE PUT METHOD SYNOPSIS

An alternate method for B<put> is available for class specific
submissions. This eliminates the need to specify TYPE and CLASS when doing a
put. The generic form of a put command using this method is shown below but
NOT detailed in the method descriptions.

  ($get,$put,$parse) = new Net::DNS::ToolKit::RR(class_type);
	or
  ($get,$put,$parse) = Net::DNS::ToolKit::RR->new(C_IN);

The generic form of a C<put> operation then becomes:

  ($newoff,@dnptrs)=$put->XYZ(\$buffer,$offset,\@dnptrs,
	$name,$ttl,$rdate,...)

The only class currently supported at this time is C_IN.

NOTE: the use of this alternate method changes the number of required
arguments to ALL put RR operations. These changes are NOT noted below in the
method descriptions.

=head1 DESCRIPTION

B<Net::DNS::ToolKit::RR> is the class loader for Resource Record classes. 
It provides an extensible wrapper for existing
classes as well as the framework to easily add new RR classes. See:
B<Net::DNS::ToolKit::RR::Template>

  From RFC 1035

  3.2.1. Format

  All RRs have the same top level format shown below:

                                    1  1  1  1  1  1
      0  1  2  3  4  5  6  7  8  9  0  1  2  3  4  5
    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
    |                      NAME                     |
    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
    |                      TYPE                     |
    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
    |                     CLASS                     |
    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
    |                      TTL                      |
    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
    |                   RDLENGTH                    |
    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--|
    |                     RDATA                     |
    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+

  NAME	an owner name, i.e., the name of the node to which this
	resource record pertains.

  TYPE	two octets containing one of the RR TYPE codes.

  CLASS	two octets containing one of the RR CLASS codes.

  TTL	a 32 bit signed integer that specifies the time interval
	that the resource record may be cached before the source
	of the information should again be consulted.  Zero
	values are interpreted to mean that the RR can only be
	used for the transaction in progress, and should not be
	cached.  For example, SOA records are always distributed
	with a zero TTL to prohibit caching.  Zero values can
	also be used for extremely volatile data.

  RDLENGTH an unsigned 16 bit integer that specifies the length
	in octets of the RDATA field.

  RDATA	a variable length string of octets that describes the
	resource.  The format of this information varies
	according to the TYPE and CLASS of the resource record.

=over 4

=item * ($get,$put,$parse) = new Net::DNS::ToolKit::RR;

Retrieves the method pointers to B<get>, B<put>, and B<parse> for Queston
section and Resource Records of a particular type.

=cut

sub new {
  my ($proto,$class) = @_;
  my $package = ref($proto) || $proto;
  my $get  = {};
  bless ($get, "${package}::get");
  my $put = ($class && ClassTxt->{$class})
	? { class => $class, } : {};
  bless ($put, "${package}::put");
  my $parse = {};
  bless ($parse, "${package}::parse");
  return ($get,$put,$parse);
}

=item * ($newoff,@common,$rdata,...) =
	$get->next(\$buffer,$offset);

Get the next Resource Record.

  input:	pointer to buffer,
		offset into buffer

  returns:	offset to next RR or section,
		(items common to all RR's)
   i.e.	$name,$type,$class,$ttl,$rdlength,
		$rdata,.... for this RR
	    or	undef if the RR is unsupported.

HERE IS THE OPPORTUNITY FOR YOU TO ADD TO THIS PACKAGE.
If your RR of interest is not supported, see:

  Net::DNS::ToolKit::RR::Template in:
  .../Net/DNS/ToolKit/Template/Template.pm

Build the support for your Resource Record and submit it to CPAN as an
extension to this package.

UN-IMPLEMENTED methods: $get->[unimplemented] returns a correct offset to
the following RR, correct @common data and a single $rdata element
containing a null ... "\0" to be precise. This works as either a numeric 0
(zero) or an end of string.

=cut

=item * ($newoff,@dnptrs)=$put->XYZ(\$buffer,$offset,\@dnptrs,
        $name,$type,$class,$ttl,$rdata,...);


Append a resource record of type XYZ to the current buffer. This is the
generic form of a B<put>.

  input:	pointer to buffer,
		offset,	[should be end of buffer]
		pointer to compressed name array,
		(items common to all RR's)
   i.e.	$name,$type,$class,$ttl,
		$rdata,.... for this RR
		in binary form if appropriate

  returns:	offset to end of RR,
		new pointer array,
	   or	empty list if the RR type is
		unsupported

  See: note above about writing new RR's

UN-IMPLEMENTED methods: $put->[unimplemented] fails miserably with a DIE
statement identifying the offending method.

=cut

=item * (@COMMON,$RDATA) = $parse->XYZ(@common,$rdata,...);

Convert non-printable and numeric data common to all records and the RR
specific B<rdata> into ascii text. In many cases this is a null
operation. i.e. for a TXT record. However, for a RR of type B<A>, the
operation would be as follows:

	EXAMPLE
Common:

  name       is already text.
  type       numeric to text
  class      numeric to text
  ttl        numeric to text
  rdlength   is a number
  rdata      RR specific conversion

Resource Record B<A> returns $rdata containing a packed IPv4 network
address. The parse operation would be:

input:

  name       foo.bar.com
  type       1
  class      1
  ttl        123
  rdlength   4
  rdata      a packed IPv4 address

output:

  name       foo.bar.com
  type       T_A
  class      C_IN
  ttl        123 # 2m 3s
  rdlength   4
  rdata      192.168.20.40

The rdata conversion is implemented internally as:

  $dotquad = inet_ntoa($networkaddress);

  where $dotquad is a printable IP address like
	192.168.20.55

UN-IMPLEMENTED methods: $parse->[unimplemented] returns correct @common
elements insofar as the type and class are present in Net::DNS::Codes.
Other elements are passed through unchanged. i.e. garbage-in, garbage-out.

=item * ($newoff,$name,type,class) =
	$get->Question(\$buffer,$offset);

  Get the Question.

  input:	pointer to buffer,
		offset
  returns:	domain name,
		question type,
		question class

=item * ($newoff,@dnptrs) =
	$put->Question(\$buffer,$offset,
	$name,$type,$class,\@dnptrs);

Append a question to the $buffer. Returns a new pointer array for compressed
names and the offset to the next RR. 

NOTE: it is up to the user to update the question count. See: L<put_qdcount>

Since the B<question> usually is the first record to be appended to the
buffer, @dnptrs may be ommitted. See the details at L<dn_comp>.

Usage: ($newoff,@dnptrs)=$put->Question(\$buffer,$offset,
	$name,$type,$class);

  input:	pointer to buffer,
		offset into buffer,
		domain name,
		question type,
		question class,
		pointer to array of
		  previously compressed names,
  returns:	offset to next record,
		updated array of offsets to
		  previous compressed names

=item * ($name,$typeTXT,$classTXT) =
	$parse->Question($name,$type,$class);

Convert non-printable and numeric data
into ascii text.

  input:	domain name,
		question type (numeric)
		question class (numeric)
  returns:	domain name,
		type TEXT,
		class TEXT

=back

=cut

1;
__END__

=head1 DEPENDENCIES

	Net::DNS::ToolKit

=head1 EXPORT

	none

=head1 AUTHOR

Michael Robinton <michael@bizsystems.com>

=head1 COPYRIGHT

    Copyright 2003 - 2011, Michael Robinton <michael@bizsystems.com>
   
All rights reserved.

This program is free software; you can redistribute it and/or modify
it under the terms of either:

  a) the GNU General Public License as published by the Free
  Software Foundation; either version 2, or (at your option) any
  later version, or

  b) the "Artistic License" which comes with this distribution.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of 
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either    
the GNU General Public License or the Artistic License for more details.

You should have received a copy of the Artistic License with this
distribution, in the file named "Artistic".  If not, I'll be glad to provide
one.

You should also have received a copy of the GNU General Public License
along with this program in the file named "Copying". If not, write to the

        Free Software Foundation, Inc.                        
        59 Temple Place, Suite 330
        Boston, MA  02111-1307, USA                                     

or visit their web page on the internet at:                      

        http://www.gnu.org/copyleft/gpl.html.

=head1 See also:

Net::DNS::Codes(3), Net::DNS::ToolKit(3), Net::DNS::ToolKit::RR::Template(3)

=cut

1;
